src/VisualStudio/Core/Def/Implementation/SolutionSize/SolutionSizeTracker.cs

```diff
@@ -49,87 +55,44 @@ public Task NewSolutionSnapshotAsync(Solution solution, CancellationToken cancel
             {
                 if (_solutionId != solution.Id)
                 {
-                    _map.Clear();
+                    Interlocked.Exchange(ref _solutionId, solution.Id);
+                    Interlocked.Exchange(ref _size, 0);
 
-                    _solutionId = solution.Id;
+                    _map.Clear();
                 }
 
                 return SpecializedTasks.EmptyTask;
             }
 
-            public async Task AnalyzeProjectAsync(Project project, bool semanticsChanged, CancellationToken cancellationToken)
-            {
-                var sum = await GetProjectSizeAsync(project, cancellationToken).ConfigureAwait(false);
-
-                _map.AddOrUpdate(project.Id, sum, (id, existing) => sum);
-
-                _size = _map.Values.Sum();
-            }
-
-            public void RemoveProject(ProjectId projectId)
+            public async Task AnalyzeSyntaxAsync(Document document, CancellationToken cancellationToken)
             {
-                long unused;
-                _map.TryRemove(projectId, out unused);
-
-                _size = _map.Values.Sum();
-            }
+                // getting tree is cheap since tree always stays in memory
+                var tree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+                var length = tree.Length;
 
-            private static async Task<long> GetProjectSizeAsync(Project project, CancellationToken cancellationToken)
-            {
-                if (project == null)
+                _map.AddOrUpdate(document.Id, id =>
```

Comment:
I don't think you can use AddOrUpdate safely because the delegates (addValueFactory and updateValueFactory) may be called more than once (if an update/add fails and has to be retried) and these two delegates have side effects (updating _size).

Unfortunately, you can't use the return value from AddOrUpdate to update _size because you need to know the old size too.

So, I think you'll have to write this out in full using TryGet / TryUpdate / TryAdd.